k8sutil/watch.py

```diff
@@ -0,0 +1,115 @@
+# returns "NamespaceList" type, then list_namespaces(watch=true) returns a stream of events with objects
+# of type "Namespace". In case this assumption is not true, user should provide return_type to Watch class's __init__.
+TYPE_LIST_SUFFIX = "List"
+
+
+class SimpleNamespace:
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+
+def _find_return_type(func):
+    for line in pydoc.getdoc(func).splitlines():
+        if line.startswith(PYDOC_RETURN_LABEL):
+            return line[len(PYDOC_RETURN_LABEL):].strip()
+    return ""
+
+
+def iter_resp_lines(resp):
+    prev = ""
+    for seg in resp.read_chunked(decode_content=False):
+        if isinstance(seg, bytes):
+            seg = seg.decode('utf8')
+        seg = prev + seg
+        lines = seg.split("\n")
+        if not seg.endswith("\n"):
+            prev = lines[-1]
+            lines = lines[:-1]
+        else:
+            prev = ""
+        for line in lines:
+            if line:
+                yield line
+
+
+class Watch(object):
+
+    def __init__(self, return_type=None):
+        self._return_type = return_type
+        self._stop = False
+
+    def stop(self):
+        self._stop = True
+
+    def stream(self, func, *args, **kwargs):
+        """Watch an API resource and stream the result back as a generators.
+
+        :param func: The API function pointer. Any parameter to the function can be passed after this parameter.
+
+        :return: Event object with these keys:
+                   'type': The type of event such as "ADDED", "DELETED", etc.
+                   'raw_object': a dict representing the watched object.
+                   'object': A model representation of raw_object. The name of model will be determined based on
+                             the func's doc string. If it cannot be determined, 'object' value will be the same
+                             as 'raw_object'.
+
+        Example:
+            v1 = k8sclient.CoreV1Api()
+            watch = k8sutil.Watch()
+            for e in watch.stream(v1.list_namespace, resource_version=1127):
+                type = e['type']
+                object = e['object']  # object is one of type return_type
+                raw_object = e['raw_object']  # raw_object is a dict
+                ...
+                if should_stop:
+                    watch.stop()
+        """
+
+        if self._return_type:
+            return_type = self._return_type
+        else:
+            return_type = _find_return_type(func)
+            if return_type.endswith(TYPE_LIST_SUFFIX):
+                return_type = return_type[:-len(TYPE_LIST_SUFFIX)]
+
+        api_client = k8sclient.ApiClient()
+        kwargs['watch'] = True
+        kwargs['_preload_content'] = False
+        resp = func(*args, **kwargs)
+        try:
+            for l in iter_resp_lines(resp):
+                j = json.loads(l)
+                j['raw_object'] = j['object']
+                if return_type:
```

Comment:
(No action required) 

Consider factoring out the marshaling so it can be isolated for testing, e.g.

```python
def marshall_obj(data, return_type):
 j = json.loads(data)
 j['raw_object'] = j['object']
 if return_type:
   ...
 return j 

def wait(...):
   ...
   try:
      for l in iter_resp_lines(resp):
         obj = self.marshall_obj(l, return_type) 
         yield obj
      ...
```